### PR TITLE
Support for the IS31FL3236 LED Driver as an RGB Matrix Driver

### PR DESCRIPTION
**Pull request by Electro707**
## Description

Adding support for the IS31FL3236 LED driver as an RGB matrix driver.

Tested with a prototype keyboard if mine with the IS31FL3236A IC (backwards compatible with the IS31FL3236 IC with the exception of a selectable PWM frequency which isn't used in this driver yet (See TODO)).

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [?] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [See TODO] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

## TODO:
- [ ] The IC supports adjustable current (from IMAX to IMAX/4). As of right now the driver drives the LEDs at IMAX. 
  -  Should we have a callable function to allow that to be changed, or is unnecessary at the moment (would be trivial to add)
- [ ] As IS31FL3236 and the IS31FL3236A are similar, should I add another define for the A version but have it use the same driver (and perhaps add the adjustable PWM for the A version as a callable function only when the A version is used)
- [x] __BUG__: On my device, after 10-20 minutes the LED driver shuts down (so all LEDs turns off) and no LED binded key-press awakes it (but the keyboard still works). Unsure if that's a QMK shutdown feature or a bug in the driver. Need to investigate further.
  - __FIXED__: I had the shutdown pin wired to the microcontroller, and I didn't set that pin to HIGH and left it floating.

### CODE REVIEW

**Review by drashna — APPROVED** [review]

File: common_features.mk [hunk 1]
```diff
@@ -286,6 +286,13 @@ endif
         QUANTUM_LIB_SRC += i2c_master.c
     endif
 
+    ifeq ($(strip $(RGB_MATRIX_DRIVER)), IS31FL3236)
+        OPT_DEFS += -DIS31FL3236 -DSTM32_I2C -DHAL_USE_I2C=TRUE
+        COMMON_VPATH += $(DRIVER_PATH)/issi
+        SRC += is31fl3236.c
+        QUANTUM_LIB_SRC += i2c_master.c
+    endif
+
```
Comment (fauxpark) [inline common_features.mk:295]: Move this before the 3731 stanza.

File: docs/feature_rgb_matrix.md [hunk 2]
```diff
@@ -108,7 +108,37 @@ const is31_led g_is31_leds[DRIVER_LED_TOTAL] = {
 Where `X_Y` is the location of the LED in the matrix defined by [the datasheet](https://www.issi.com/WW/pdf/31FL3733.pdf) and the header file `drivers/issi/is31fl3733.h`. The `driver` is the index of the driver you defined in your `config.h` (Only `0` right now).
 
 ---
+# IS31FL3236 :id=is31fl3236
```
Comment (fauxpark) [inline docs/feature_rgb_matrix.md:111]: As above.

File: quantum/rgb_matrix.h [hunk 3]
```diff
@@ -33,6 +33,8 @@
 #    include "is31fl3737.h"
 #elif defined(IS31FL3741)
 #    include "is31fl3741.h"
+#elif defined(IS31FL3236)
```
Comment (fauxpark) [inline quantum/rgb_matrix.h:36]: etc.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: common_features.mk (hunk 1 above)
Comment (Electro707) [reply common_features.mk]: Should be moved by recent commit

File: docs/feature_rgb_matrix.md (hunk 2 above)
Comment (Electro707) [reply docs/feature_rgb_matrix.md]: Same as above

File: quantum/rgb_matrix.h (hunk 3 above)
Comment (Electro707) [reply quantum/rgb_matrix.h]: Same as above^2

**Comment by Electro707** [comment]
This PR should be ready to merge, all of the review's comments have been addressed last time I checked.

**Comment by Electro707** [comment]
@drashna @fauxpark Do you want to re-open this PR? I know it's been a while, but I do eventually want to merge this.
I will look into re-basing master on this branch to see if there are any new changes to the LED drive that this PR conflicts with since this PR was created.

**Comment by tzarc** [comment]
> I will look into re-basing master on this branch 
Just an FYI, core changes need to go to `develop`, not `master`.

**Comment by Electro707** [comment]
Thanks for also changing this PR's merging branch. I think the develop branch is something relatively new, at least since I originally created this PR.

**Comment by Electro707** [comment]
Fixed some issues that occurred when re-basing. Tested with my keypad.
Fixed docs and added the `g_is31_leds` LED definition instead of amusing a sequential connection.
This PR should be all set, at least from my end

File: drivers/led/issi/is31fl3236.c [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+void IS31FL3236_set_color(int index, uint8_t red, uint8_t green, uint8_t blue) {
+    is31_led led;
+    memcpy_P(&led, (&g_is31_leds[index]), sizeof(led));
+
+    g_pwm_buffer[led.r]  = red;
+    g_pwm_buffer[led.g]  = green;
+    g_pwm_buffer[led.b]  = blue;
+    g_pwm_buffer_update_required = true;
```
Comment (drashna) [inline drivers/led/issi/is31fl3236.c:100]:
```suggestion
    g_pwm_buffer[led.r]          = red;
    g_pwm_buffer[led.g]          = green;
    g_pwm_buffer[led.b]          = blue;
    g_pwm_buffer_update_required = true;
```

File: drivers/led/issi/is31fl3236.c [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+void IS31FL3236_set_led_control_register(uint8_t index, bool red, bool green, bool blue) {
+    is31_led led;
+    memcpy_P(&led, (&g_is31_leds[index]), sizeof(led));
+
+    g_led_control_registers[led.r] = red;
+    g_led_control_registers[led.g] = green;
+    g_led_control_registers[led.b] = blue;
+    g_led_control_update_required = true;
```
Comment (drashna) [inline drivers/led/issi/is31fl3236.c:116]:
```suggestion
    g_led_control_update_required  = true;
```

**Review by drashna — CHANGES_REQUESTED** [review]
Lint no happy.

File: drivers/led/issi/is31fl3236.c (hunk 4 above)
Comment (Electro707) [reply drivers/led/issi/is31fl3236.c]: Should be fixed in 133d473

File: drivers/led/issi/is31fl3236.c (hunk 5 above)
Comment (Electro707) [reply drivers/led/issi/is31fl3236.c]: Should be fixed in 133d473

**Review by drashna — APPROVED** [review]

**Comment by Electro707** [comment]
@drashna @fauxpark Another request to re-open this PR to merge it, as the review request didn't happen last time

**Pull request closed without merging**